### PR TITLE
Add boost to KerrSchild

### PR DESCRIPTION
**Pull request by guilara**
## Proposed changes

Add a boost to the KerrSchild solution. This is done by boosting the coordinates, null form and derivatives of the H function defining the KerrSchild solution. (For the case when there is a boost the solution depends on time as well, but we don't code up the time dependence of the coordinates for simplicity.)

Evolving this solution will require time dependent translation maps to keep the excision surface inside the apparent horizon (#5576)

This reuses items from https://github.com/sxs-collaboration/spectre/compare/develop...nilsvu:spectre:kerr_boost

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp [hunk 1]
```diff
@@ -214,6 +214,16 @@ namespace Solutions {
  *
  * Right now we use (\f$\ref{eq:sphertocartsimple}\f$), but we may
  * wish to use the other transformation in the future.
+ *
+ * ## Boost of the Kerr-Schild solution
+ *
+ * We add initial momentum to the solution by applying a Lorentz boost to the
+ * metric. Given the form of the Kerr-Schild metric in terms of a covariant
```
Comment (nikwit) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: Can you rephrase this to something more readable? Perhaps "The Kerr-Schild metric can be expressed covariantly in terms of (...). We construct the boosted metric by boosting these object individually."

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp [hunk 2]
```diff
@@ -368,143 +406,175 @@ class KerrSchild : public AnalyticSolution<3_st>,
...
+        const;
+
+    void operator()(
+        gsl::not_null<tnsr::I<DataType, 3, Frame>*> x_minus_center_boosted,
+        gsl::not_null<CachedBuffer*> /*cache*/,
         internal_tags::x_minus_center<DataType, Frame> /*meta*/) const;
 
-    void operator()(const gsl::not_null<Scalar<DataType>*> a_dot_x,
-                    const gsl::not_null<CachedBuffer*> cache,
+    void operator()(gsl::not_null<Scalar<DataType>*> a_dot_x,
```
Comment (nikwit) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:434]: why did you delete all these `const` specifiers?

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.cpp [hunk 3]
```diff
@@ -391,8 +450,7 @@ void KerrSchild::IntermediateComputer<DataType, Frame>::operator()(
   const auto& lapse = get(cache->get_var(*this, gr::Tags::Lapse<DataType>{}));
   const auto& lapse_squared =
       get(cache->get_var(*this, internal_tags::lapse_squared<DataType>{}));
-  get(*deriv_lapse_multiplier) =
-      -square(null_vector_0_) * lapse * lapse_squared;
+  get(*deriv_lapse_multiplier) = -lapse * lapse_squared;
```
Comment (nikwit) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.cpp:457]: why can you set `square(null_vector_0_)` to 1 here but not other places?

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp [hunk 4]
```diff
@@ -58,6 +59,80 @@ void lorentz_boost_matrix(
     (*boost_matrix).get(i + 1, i + 1) += 1.0;
   }
 }
+
+template <size_t SpatialDim>
+tnsr::Ab<double, SpatialDim, Frame::NoFrame> lorentz_boost_matrix(
+    const std::array<double, SpatialDim>& velocity) {
+  tnsr::I<double, SpatialDim, Frame::NoFrame> velocity_as_tensor{};
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp]: `tnsr::I<double, SpatialDim, Frame::NoFrame> velocity_as_tensor(velocity)` should work I think

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp [hunk 5]
```diff
@@ -53,5 +56,51 @@ template <size_t SpatialDim>
...
+template <size_t SpatialDim>
+tnsr::Ab<double, SpatialDim, Frame::NoFrame> lorentz_boost_matrix(
+    const std::array<double, SpatialDim>& velocity);
+/// @}
+
+/// @{
+/*!
+ * \ingroup SpecialRelativityGroup
+ * \brief Apply a Lorentz boost to the spatial part of a one form.
+ * \details The zero component of the one form is also specified.
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: Maybe "This requires passing the 0th component of the one form as an additional argument.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp [hunk 6]
```diff
@@ -53,5 +56,51 @@ template <size_t SpatialDim>
...
+    const double vector_component_0,
+    const std::array<double, SpatialDim>& velocity);
+/// @}
+
+/// @{
+/*!
+ * \ingroup SpecialRelativityGroup
+ * \brief Apply a Lorentz boost to a one form.
+ *
+ * Note that the Lorentz matrix is symmetric.
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: I don't think this is relevant for this function.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp [hunk 7]
```diff
@@ -53,5 +56,51 @@ template <size_t SpatialDim>
...
+void lorentz_boost(
+    const gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> result,
+    const tnsr::a<DataType, SpatialDim, Frame>& vector,
+    const std::array<double, SpatialDim>& velocity);
+/// @}
+
+/// @{
+/*!
+ * \ingroup SpecialRelativityGroup
+ * \brief Apply a Lorentz boosts for each index of a rank-2 tensor with
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: maybe something like: "(...) boost to each component"

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp [hunk 8]
```diff
@@ -53,5 +56,51 @@ template <size_t SpatialDim>
...
+
+template <size_t SpatialDim>
+tnsr::Ab<double, SpatialDim, Frame::NoFrame> lorentz_boost_matrix(
+    const std::array<double, SpatialDim>& velocity);
+/// @}
+
+/// @{
+/*!
+ * \ingroup SpecialRelativityGroup
+ * \brief Apply a Lorentz boost to the spatial part of a one form.
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: You mean vector here instead of one form, right?

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp [hunk 9]
```diff
@@ -53,5 +56,51 @@ template <size_t SpatialDim>
...
+/*!
+ * \ingroup SpecialRelativityGroup
+ * \brief Apply a Lorentz boost to a one form.
+ *
+ * Note that the Lorentz matrix is symmetric.
+ */
+template <typename DataType, size_t SpatialDim, typename Frame>
+void lorentz_boost(
+    const gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> result,
+    const tnsr::a<DataType, SpatialDim, Frame>& vector,
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: one_form

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp [hunk 10]
```diff
@@ -58,6 +59,80 @@ void lorentz_boost_matrix(
...
+  for (size_t i = 0; i < SpatialDim + 1; ++i) {
+    result->get(i) = 0.;
+    for (size_t j = 0; j < SpatialDim + 1; ++j) {
+      result->get(i) += boost_matrix.get(i, j) * vector.get(j);
+    }
+  }
+}
+
+template <typename DataType, size_t SpatialDim, typename Frame>
+void lorentz_boost(
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp:109]: is there a sensible situation where different indices of a tensor would be boosted with different velocities? Is this generalization necessary?

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp [hunk 11]
```diff
@@ -58,6 +59,80 @@ void lorentz_boost_matrix(
     (*boost_matrix).get(i + 1, i + 1) += 1.0;
   }
 }
+
+template <size_t SpatialDim>
+tnsr::Ab<double, SpatialDim, Frame::NoFrame> lorentz_boost_matrix(
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp:64]: can you make this a symmetric tensor?

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp [hunk 12]
```diff
@@ -58,6 +59,80 @@ void lorentz_boost_matrix(
...
+    const std::array<double, SpatialDim>& velocity_first_index,
+    const std::array<double, SpatialDim>& velocity_second_index) {
+  if (velocity_first_index == make_array<SpatialDim>(0.) and
+      velocity_second_index == make_array<SpatialDim>(0.)) {
+    *result = tensor;
+    return;
+  }
+  const auto boost_matrix_1 = lorentz_boost_matrix(velocity_first_index);
+  const auto boost_matrix_2 = lorentz_boost_matrix(velocity_second_index);
+  for (size_t i = 0; i < SpatialDim + 1; ++i) {
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp:118]: [optional] `tenex::evaluate<ti::a, ti::b>(result, boost_matrix_1(ti::C, ti::a) * boost_matrix_2(ti::D, ti::b) * tensor(ti::c, ti::d)` is quite concise.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp [hunk 13]
```diff
@@ -53,5 +56,51 @@ template <size_t SpatialDim>
...
+
+/// @{
+/*!
+ * \ingroup SpecialRelativityGroup
+ * \brief Apply a Lorentz boost to a one form.
+ *
+ * Note that the Lorentz matrix is symmetric.
+ */
+template <typename DataType, size_t SpatialDim, typename Frame>
+void lorentz_boost(
```
Comment (nikwit) [inline src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp:82]: Ideally, there would be one implementation that works for a generic tensor type so do not have all of these overloads. If you think that would be overkill for now, maybe just add a comment that this can be done in the future.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrappedGr.cpp [hunk 14]
```diff
@@ -44,7 +44,7 @@ void compare_different_wrapped_solutions(const double mass,
 template <typename SolutionType>
 void test_copy_and_move(const SolutionType& solution) {
   test_copy_semantics(solution);
-  auto solution_copy = solution;
+  const auto& solution_copy = solution;
```
Comment (nikwit) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrappedGr.cpp]: this is just a reference now. Can you just pass in `solution` directly?

File: tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp [hunk 15]
```diff
@@ -426,5 +426,55 @@ void verify_consistency(const Solution& solution, const double time,
                                               derivative_delta),
       derivative_approx);
 }
+
+/// Check the consistency of dependent quantities returned by a
```
Comment (nikwit) [inline tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp]: dependent on what?

**Review by nikwit — CHANGES_REQUESTED** [review]
nice work overall. Two points:

1.) I think we discussed that with a boost, `KerrSchild` is technically time dependent but you have only implemented it for `t=0`. It is therefore technically no longer a `AnalyticSolution`.  Can you make sure to document this?

2.) As we discussed, could you write a test that computes tensors like the extrinsic curvature and the spacetime metric once boosted and once unboosted  and make sure they correspond to a manual transformation?

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 2 above)
Comment (guilara) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:434]: clang-tidy was complaining. Said only has effect in the cpp

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.cpp (hunk 3 above)
Comment (guilara) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.cpp:457]: I don't think its set to one, just that it seems not in the formula for the spatial derivative of the lapse.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp (hunk 11 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp:64]: Can I? This has one upper and one lower index

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 1 above)
Comment (guilara) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: done

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp (hunk 4 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp]: Done. Seems to work

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 5 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: Done.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 13 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp:82]: I think I wouldn't want to delay this anymore. Added comment.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrappedGr.cpp (hunk 14 above)
Comment (guilara) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_WrappedGr.cpp]: Done.

File: tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp (hunk 15 above)
Comment (guilara) [reply tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp]: On other metric quantities. Changed doc.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 8 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: One form. This is because here and in SpEC they chose (rather confusingly) the lorentz boost matrix to be the usual _inverse_ lorentz boost matrix (or that for one forms).

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 9 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: Done

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 7 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: Okay

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 6 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: Okay. Removed comment

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp (hunk 10 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp:109]: Also inherited from the implementation in SpEC, maybe to save overloads when boosting tensors with mixed up/down indices. Then one just changes the sign of the velocity in the relevant entry.

**Comment by guilara** [comment]
@nikwit I've added one more test to manually boost the spacetime metric and check that I get the same as the implementation. Not entirely sure how to do it for the extrinsic curvature. I think I need to upgrade it to 4d tensor (not sure about this) then boost it and compare a 4d tensor obtained from the implementation.

**Comment by guilara** [comment]
@nikwit was suggesting moving the KerrBoost to AnalyticData instead of AnalyticSolutions while I suggest leaving it here and add a warning in the doc.  

The reasoning is mainly because the Kerr boost as it is currently implemented (and as it is implemented in SpEC) is only usable for initial data and some assumptions are being made. For instance, we are not coding up the time dependence at all (the solution now is both function of position and time since its moving, but we set t = 0), and assumptions on the time derivative of the lapse and shift (which will be rewritten anyway when evolved).  

@nilsvu any thoughts on this?

**Comment by nilsvu** [comment]
@geoffrey4444 Could you review this one too, since you wrote the Lorentz transformations here and boosts in SpEC?

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp [hunk 16]
```diff
@@ -214,6 +214,28 @@ namespace Solutions {
  *
  * Right now we use (\f$\ref{eq:sphertocartsimple}\f$), but we may
  * wish to use the other transformation in the future.
+ *
+ * ## Boost of the Kerr-Schild solution
+ *
+ * We add initial momentum to the solution by applying a Lorentz boost to the
+ * metric. Since the Kerr-Schild metric can be expressed covariantly in terms of
+ * the Minkowski metric, a scalar function and a one form, we constructing the
+ * metric by boosting the these objects individually. Notice that we also need
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: <s>the</s>

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp [hunk 17]
```diff
@@ -214,6 +214,28 @@ namespace Solutions {
...
+ * to appropriately boost the coordinates to the boosted frame.
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: I believe we're evaluating the boosted Kerr-Schild metric in the boosted frame, meaning that in these coordinates the black hole does not move, correct? So we inverse-boost the input coordinates to get to the unboosted frame, evaluate Kerr-Schild quantities in the unboosted frame, then boost these quantities. Please clarify the docs here.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp [hunk 18]
```diff
@@ -214,6 +214,28 @@ namespace Solutions {
...
+ *
+ * \warning While technically the boosted Kerr-Schild metric is dependent on
+ * both the time and space coordinates, we have implemented it only at $t = 0$
+ * as in SpEC. Therefore it is technically not an analytic solution and should
+ * not be used to compute errors with respect to it.
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:231]: The solution should be stationary in the boosted frame; so shouldn't all time derivatives be zero and this be a solution at any $t$? Am I misunderstanding something?

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp (hunk 11 above)
Comment (nilsvu) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.cpp:64]: Not currently supported.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 8 above)
Comment (nilsvu) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: The argument type is a vector (`tnsr::I`), so I also think this documentation should read `vector`. Refer to the docs of `lorentz_boost_matrix` above for details and math. I also find those docs confusing though. I believe they say that $\Lambda^a_\bar{a}$ is computed, which boosts a vector from barred to unbarred or a one-form from unbarred to barred (consistent with Blandford/Thorne Sec. 2.7). So my suggestion is:

- The docs and argument for this function should read `vector` (they boost a vector)
- The function for the one-form below should apply the inverse Lorentz transformation (by inverting the sign of the velocity). This means you add a minus sign to the function below and remove the minus signs from where the function is used in Kerr-Schild. Make sure to clarify this in the docs of the function.

@geoffrey4444 can you confirm or clarify this?

File: tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp [hunk 19]
```diff
@@ -426,5 +426,55 @@ void verify_consistency(const Solution& solution, const double time,
                                               derivative_delta),
       derivative_approx);
 }
+
+/// Check the consistency of quantities depenent on other metric quantities
```
Comment (nilsvu) [inline tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp]: typo: dependent

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp [hunk 20]
```diff
@@ -366,13 +380,116 @@ void test_zero_spin_optimization(const DataType& used_for_size) {
       });
 }
 
+template <typename Frame, typename DataType>
+void test_boosted_for_zero_velocity(const DataType& used_for_size) {
+  const std::array<double, 3> boost_velocity{{0.0, 0.0, 1.0e-50}};
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp]: Maybe `std::numeric_limits<double>::epsilon()`?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp [hunk 21]
```diff
@@ -366,13 +380,116 @@ void test_zero_spin_optimization(const DataType& used_for_size) {
       });
 }
 
+template <typename Frame, typename DataType>
+void test_boosted_for_zero_velocity(const DataType& used_for_size) {
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp:396]: Nice test!

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp [hunk 22]
```diff
@@ -366,13 +380,116 @@ void test_zero_spin_optimization(const DataType& used_for_size) {
...
+  const std::array<double, 3> spin{{0.0, 0.0, 0.0}};
+  const std::array<double, 3> center{{0.0, 0.0, 0.0}};
+  gr::Solutions::KerrSchild solution(mass, spin, center);
+
+  // Boosted KerrSchild solution
+  const std::array<double, 3> velocity{{0.5, 0.0, 0.0}};
+  gr::Solutions::KerrSchild boosted_solution(mass, spin, center, velocity);
+
+  // Need to evaluate at the boosted coordinates
+  // Boost coordinates
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp:445]: These are the unboosted coords, no? (related to comments above)

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp [hunk 23]
```diff
@@ -366,13 +380,116 @@ void test_zero_spin_optimization(const DataType& used_for_size) {
...
+      [&all_tags_zero_velocity, &all_tags_tiny_velocity](auto tag_v) {
+        using tag = tmpl::type_from<decltype(tag_v)>;
+        CHECK_ITERABLE_APPROX(get<tag>(all_tags_zero_velocity),
+                              get<tag>(all_tags_tiny_velocity));
+      });
+}
+
+template <typename Frame, typename DataType>
+void test_boosted_spacetime_metric(const DataType& used_for_size) {
+  // Test that the extrinsic curvature of the boosted solution indeed
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp]: extrinsic curvature -> spacetime metric? Also test extrinsic curvature?

File: tests/Unit/PointwiseFunctions/SpecialRelativity/Test_LorentzBoostMatrix.cpp [hunk 24]
```diff
@@ -86,3 +167,9 @@ SPECTRE_TEST_CASE(
   d = large_velocity_squared;
   CHECK_FOR_DOUBLES(test_lorentz_boost_matrix_analytic, (1, 2, 3));
 }
+
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.SpecialRelativity.LorentzBoost",
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/SpecialRelativity/Test_LorentzBoostMatrix.cpp]: combine with test case above

**Review by nilsvu — CHANGES_REQUESTED** [review]
Yes please keep this class in `AnalyticSolution` with the warning in the docs.

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 8 above)
Comment (guilara) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: You are right Nils. My mistake here, Niko. It is the transformation for a vector. But I should take the inverse Lorentz matrix (or add a sign to the velocity) as they transform with the inverse matrix wrt one forms.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 17 above)
Comment (guilara) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: Will clarify. I got confused about it myself at some point. We start with coordinates in the frame where the black hole is moving. Then boost them to the frame where the black hole is at rest, and where we can construct the stationary solution. Then we boost all covariant quantities back to the frame where it is moving.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 18 above)
Comment (guilara) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:231]: I believe it is stationary in the rest frame, which is not the frame we want. In the frame we want, the black hole is moving and so the metric will be dependent on t and x. I didn't code this t dependence  since I didn't see this in SpEC --it seems to me it should only enter in the function that computes the radius to the center. 

I also didn't code the time derivatives of the lapse and shift in the frame where the black hole is moving (which are zero in the rest frame of course) but left them as zero. This should be fine because I think they are replaced by the gauge functions we use to evolve this.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp (hunk 23 above)
Comment (guilara) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp]: Not yet sure how to do it. I think I would need to elevate the extrinsic curvature to a 4d tensor in both frames.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp (hunk 22 above)
Comment (guilara) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp:445]: Maybe the terminology ''boosted/unboosted'' is a bit confusing.  Rest frame (of the BH) coords I would say.

File: tests/Unit/PointwiseFunctions/SpecialRelativity/Test_LorentzBoostMatrix.cpp (hunk 24 above)
Comment (guilara) [reply tests/Unit/PointwiseFunctions/SpecialRelativity/Test_LorentzBoostMatrix.cpp]: Combined

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp (hunk 20 above)
Comment (guilara) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_KerrSchild.cpp]: Okay

File: tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp (hunk 19 above)
Comment (guilara) [reply tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/VerifyGrSolution.hpp]: Corrected

File: src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp (hunk 8 above)
Comment (nilsvu) [reply src/PointwiseFunctions/SpecialRelativity/LorentzBoostMatrix.hpp]: I think the sign for this function is correct, but the one below (for one-forms) should have a minus sign.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 17 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: Hm, for initial data I thought we're solving in a frame where the BHs are at rest.. @geoffrey4444 can you clarify how this works in SpEC?

**Comment by geoffrey4444** [comment]
> @geoffrey4444 Could you review this one too, since you wrote the Lorentz transformations here and boosts in SpEC?

I'll try but really swamped this week.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 17 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp]: Talked about this in the meeting today. I think you're correct. For initial data we go into a rotating frame that gets rid of the translation but keeps the Lorentz contraction.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp (hunk 18 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:231]: Ok I think you're correct. Let me look at the SpEC implementation too.

**Comment by guilara** [comment]
@nilsvu  I have rebased this  and added an assert that would throw an error if we call a tag I can't compute for the boost.

**Review by geoffrey4444 — APPROVED** [review]
@nilsvu does this look ready to go to you?

**Review by nilsvu — APPROVED** [review]
Looks good. Thanks for fixup this up. Pleas squash your commits. Also to keep the git blame record straight, could you add this to your commit message:

```sh
# Two empty lines


Co-authored-by: Nils Vu <nilsvu@caltech.edu>
```

**Comment by nilsvu** [comment]
It's two empty lines, no `>` symbols

**Comment by guilara** [comment]
> It's two empty lines, no `>` symbols

Oh, sure

**Comment by nilsvu** [comment]
Thank you for getting this done @guilara !

**Comment by guilara** [comment]
> Thank you for getting this done @guilara !

Thank you for all the detailed help with it
